src/Shared/Microsoft.DotNet.Internal.DependencyInjection/RefCountedObject.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Threading;
+
+namespace Microsoft.DotNet.Internal.DependencyInjection
+{
+    public sealed class RefCountedObject<T>
+    {
+        private int _refCount;
+
+        public RefCountedObject(T value)
+        {
+            _refCount = 0;
+            Value = value;
+        }
+
+        public T Value { get; }
+
+        public void AddRef()
+        {
+            Interlocked.Increment(ref _refCount);
+        }
+
+        public void Release()
```

Comment:
Should it have a "release async"?

Reply:
(That the Reference.DisposeAsync could call?)

Reply:
I don't think that would be all that useful. We still have to implement Dispose because some things don't play nicely with DisposeAsync, so we would still have the .GetResult() stuff.